### DISCUSSION TITLE
Set up multiple auth methods on a route, expecting at least one of them to pass

### DISCUSSION THREAD

**Discussion by pawamoy** [fastapi/fastapi#5357] [Questions] [answered]
Hello, I'm having trouble setting two authentication methods on a route.

Use-case: the operation behind this route can be very destructive. I want to protect it so that:

- only admins can use it
- or a particular service account can use it

The admins use a cookie based authentication method, while the service account uses a basic auth method.
Until now I was raising HTTPException with a 401 Unauthorized directly from my `is_admin` or `is_service_account` dependencies.

That means I cannot create a dependency that mixes them both:

```python
def is_admin_or_svc_account(
    admin = Depends(is_admin),
    svc_account = Depends(is_service_account),
): ...
```

...since it expects both of them to pass instead of just one (one of them failing equals exception raised).

To make it work, I had to modify `is_admin` or `is_service_account` dependencies so that they don't raise an HTTPException directly but rather return a boolean. That's not ideal as it complicates the code a lot: I have to catch downstream exceptions, return and check `None` in other places, instead of just letting exceptions bubbling up.

Is there a simple way to setup "fallback" authentication methods on a route?
Or a way to inject "one of" several dependencies, using the first one that doesn't raise an exception for example?

```python
@app.post("/dangerous", dependencies=[Depends(one_of(is_admin, is_service_account))])
```

**Comment by Jedore** [comment]
I don't think there is a simple way without modifying your code.

**Reply by pawamoy** [reply]
I see, thanks for your answer :slightly_smiling_face:

**Reply by tiangolo** [reply, marked as answer]
You would need to do what you did.

An option is to have a `optional_is_admin` that doesn't raise and use it inside of `is_admin` that checks it and raises.

Your code would keep mostly the same except for the one that accepts any of the possible authentication methods, that one would use the `optional_is_admin`, check both, and then raise if no auth passed.

**Reply by pawamoy** [reply]
I see, makes sense. I also did try to implement such a `one_of` utility, but failed miserably. Anyway, I don't need this anymore 🙂 Thanks!
